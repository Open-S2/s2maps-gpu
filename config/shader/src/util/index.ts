export * from './bundle';
export * from './tree';
