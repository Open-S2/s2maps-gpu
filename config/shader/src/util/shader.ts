import { Tree } from '@lezer/common';
import { ASTParser, VirtualTable, SymbolTableT, ParsedBundle, ParsedModule, ParsedModuleCache, CompressedNode } from '../types';
import { formatMurmur53, toMurmur53 } from './hash';
import { PREFIX_VIRTUAL } from '../constants';

const EMPTY_LIST = [] as any[];
const EMPTY_TABLE = {} as any;

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const makeLoadModule = <T extends SymbolTableT = any>(
  parseShader: (code: string) => Tree,
  makeASTParser: (code: string, tree: Tree, name?: string) => ASTParser<T>,
  compressAST: (code: string, tree: Tree, symbols?: string[]) => CompressedNode[],
  decompressAST: (nodes: CompressedNode[], symbols?: string[]) => Tree,
) => (
  code: string,
  name: string = 'main',
  entry?: string,
  compressed: boolean = false,
): ParsedModule => {
  if (code == null) throw new Error(`Shader code ${name} undefined`);
  if (typeof code !== 'string') throw new Error(`Shader code ${name} is not a string`);
  let tree = parseShader(code);

  const astParser = makeASTParser(code, tree, name);
  const table = astParser.getSymbolTable();
  const shake = astParser.getShakeTable(table);

  if (compressed) {
    const {symbols} = table;
    tree = decompressAST(compressAST(code, tree, symbols), symbols);
  }
  const hash = toMurmur53(code);

  return bindEntryPoint({name, code, hash, table, shake, tree}, entry);
}

// Use cache to load modules
export const makeLoadModuleWithCache = (
  loadModule: (code: string, name?: string, entry?: string, compressed?: boolean) => ParsedModule,
  defaultCache: ParsedModuleCache,
) => (
  code: string,
  name?: string,
  entry?: string,
  cache: ParsedModuleCache | null = defaultCache,
): ParsedModule => {
  if (!cache) return loadModule(code, name, entry, true);

  const hash = toMurmur53(code);
  const cached = cache.get(hash);
  if (cached) {
    return bindEntryPoint(cached, entry);
  }
  
  const module = loadModule(code, name, undefined, true);
  cache.set(hash, module);
  return bindEntryPoint(module, entry);
}

// Load a static (inert) module
export const loadStaticModule = (code: string, name: string, entry?: string) => {
  const hash = toMurmur53([code, entry]);
  return ({ name, code, hash, table: EMPTY_TABLE });
}

// Load a virtual (generated) module
export const loadVirtualModule = <T extends SymbolTableT = any>(
  virtual: VirtualTable,
  initTable: Partial<T> = EMPTY_TABLE,
  entry?: string,
  hash?: number,
  code?: string,
  key?: number,
) => {
  let symbols = initTable.symbols ?? EMPTY_LIST;

  code = code ?? `@virtual [${symbols.join(' ')}]`;
  hash = hash ?? toMurmur53(code);
  key  = key  ?? hash;

  const name = `${PREFIX_VIRTUAL}${formatMurmur53(key).slice(0, 6)}`;

  const table = {
    symbols,
    visibles: symbols,
    ...initTable,
  };
  return { name, code, hash, table, entry, virtual, key };
}

// Set entry point of a module, returns new bundle/module.
// Is the same instance as the original (key = old key/hash), so it merges with copies of itself.
// But is structurally different (hash = new key), so differences in links are reflected in the shader hash.
export const bindEntryPoint = <T extends ParsedBundle | ParsedModule>(bundle: T, entry?: string): T => {
  let {key, hash, module, table} = bundle as any;

  table = table ?? module?.table;
  hash = hash ?? module?.hash;

  if (entry == null && table.symbols?.includes('main')) entry = 'main';
  if (entry == null) return bundle;

  const structural = toMurmur53([hash, entry]);
  return {...bundle, entry, hash: structural, key: key ?? hash};
};
