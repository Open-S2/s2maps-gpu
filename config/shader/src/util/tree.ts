import { SyntaxNode, Tree } from '@lezer/common';
import { CompressedNode } from '../types';

// Gather all child nodes (may be implicit)
export const getChildNodes = (node: SyntaxNode) => {
  const out = [] as SyntaxNode[];
  let n = node.firstChild;
  while (n) {
    out.push(n);
    n = n.nextSibling;
  }
  return out;
}

// Look for error nodes in a tree
export const hasErrorNode = (tree: Tree) => {
  const cursor = tree.cursor();
  do {
    const {type, from, to} = cursor;
    if (type.isError) return cursor.node;
  } while (cursor.next());
  return false;
}

// AST node to tree view + side-by-side tokens
export const formatAST = (node: SyntaxNode, code?: string, depth: number = 0) => {
  const {type, from ,to} = node;
  const prefix = '  '.repeat(depth);

  let child = node.firstChild;

  const text = code != null ? code.slice(node.from, node.to).replace(/\n/g, "⮐ ") : '';
  let out = [] as string[];

  let line = `${prefix}${type.name}`;
  const n = line.length;
  line += ' '.repeat(60 - n);
  line += text;
  out.push(line);

  while (child) {
    out.push(formatAST(child, code, depth + 1));
    child = child.nextSibling;
  }
  return out.join("\n");
}

// AST node to S-expression
export const formatASTNode = (node: SyntaxNode) => {
  const {type} = node;
  let child = node.firstChild;
  let inner = [] as string[];
  while (child) {
    inner.push(formatASTNode(child));
    child = child.nextSibling;
  }
  const space = inner.length ? ' ' : '';
  return `(${type.name}${space}${inner.join(" ")})`;
}

const getOpsMap = (ops: string[]) => {
  const opToIndex = new Map<string, number>();
  const indexToOp = new Map<number, string>();
  ops.forEach((op, i) => {
    opToIndex.set(op, i);
    indexToOp.set(i, op);
  });
  return {opToIndex, indexToOp};
};

// Compress AST node
export const makeASTEmitter = (
  out: any[],
  ops: string[],
  symbols: string[] = [],
) => {
  let offset = 0;
  const encode = (x: number) => x - offset;
  const {opToIndex} = getOpsMap(ops);

  return (
    type: string,
    from: number,
    to: number,
    arg?: any,
  ) => {

    const i = opToIndex.get(type)!;
    const row = [i, encode(from), encode(to)];
    if (arg != null) row.push(symbols.indexOf(arg));

    offset = from;
    out.push(row);
  };
};

// Decompress a compressed AST on the fly by returning a pseudo-tree-cursor.
export const makeASTDecompressor = (
  ops: string[]
) => (
  nodes: CompressedNode[],
  symbols: string[] = [],
): Tree => {
  const {indexToOp} = getOpsMap(ops);

  const tree = {
    __nodes: nodes,
    cursor: () => {
      let offset = 0;
      const decode = (d: number) => d + offset;

      let i = -1;
      const n = nodes.length;

      const next = () => {
        const hasNext = ++i < n;
        if (!hasNext) return false;

        const node = nodes[i];
        const [op, d1, d2, arg] = node;

        self.type.name = indexToOp.get(op)!;
        self.from = decode(d1);
        self.to = decode(d2);
        self.arg = arg != null ? symbols[arg] : null;
        offset = self.from;

        return true;
      };

      const lastChild = () => {
        const {to} = self;
        do {
          const node = nodes[i + 1];
          if (node && decode(node[1]) >= to) return false;
        } while (next());
        return false;
      }

      const self = {
        type: {name: ''},
        node: {parent: {type: {name: 'Program'}}},
        from: 0,
        to: 0,
        arg: undefined,
        next,
        lastChild,
      } as any;

      next();

      return self;
    },
  } as any as Tree;
  return tree;
}
