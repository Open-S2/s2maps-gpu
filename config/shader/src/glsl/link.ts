import { Tree } from '@lezer/common';

import { defineConstants, defineEnables, loadModule, loadModuleWithCache, DEFAULT_CACHE } from './shader';
import { rewriteUsingAST } from './ast';
import { makeLinker, makeLinkCode, makeLinkBundle, makeLinkModule } from '../util/link';

import { GLSL_VERSION } from './constants';

// Override GLSL version/prefix
let PREAMBLE = `#version ${GLSL_VERSION}`;
export const setPreamble = (s: string): string => PREAMBLE = s;
export const getPreamble = (): string => PREAMBLE;
const getPreambles = (): string[] => [getPreamble()];

// No preprocessor renames
const getRenames = () => new Map<string, string>();

export const linker     = makeLinker(getPreambles, getRenames, defineConstants, defineEnables, rewriteUsingAST);
export const linkBundle = makeLinkBundle(linker);
export const linkModule = makeLinkModule(linker);
export const linkCode   = makeLinkCode(linker, loadModuleWithCache, DEFAULT_CACHE);
