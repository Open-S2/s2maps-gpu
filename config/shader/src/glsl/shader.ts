import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { parser } from './grammar/glsl';
import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse GLSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make GLSL definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

/** Make GLSL enable definitions */
export const defineEnables = (enabled: string[]) => enabled.map(e => `#extension ${e}â€‹ : enable`).join('\n');

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);

// Simple whitespace removal
const minifyCode = (code: string) => {
  code = code.replace(/ +/g, ' ');
  code = code.replace(/\n+/g, '\n');
  return code;
};

/** ES/CommonJS Transpiler */
export const transpileGLSL = makeTranspile('glsl', 'glsl', loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
glsl`...`
``` */
export const glsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as GLSL `float` */
export const float = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : 0;  
};
/** Format `number` as GLSL `uint` */
export const uint = (x: number) => Math.round(x).toString();
/** Format `number` as GLSL `int` */
export const int = (x: number) => Math.round(x).toString();
