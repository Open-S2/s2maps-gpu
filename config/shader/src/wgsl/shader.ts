import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { removeComments, removeWhiteSpace, renameLocals } from './minify';
import { parser } from './grammar/wgsl';

import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse WGSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make WGSL constant definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (k[0] !== '@' && defs[k] != null) out.push(`const ${k} = ${defs[k]};`);
  return out.join("\n");
}

/** Make WGSL constant definitions */
export const defineEnables = (enabled: string[]) => enabled.length ? `enable ${enabled.join(', ')};` : '';

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);

// Simple whitespace / comment removal
const minifyCode = (code: string) => {
  code = removeComments(code);
  code = renameLocals(code);
  code = removeWhiteSpace(code);
  return code;
};

/** ES/CommonJS Transpiler */
export const transpileWGSL = makeTranspile('wgsl', 'wgsl', loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
wgsl`...`
``` */
export const wgsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as WGSL `f32` */
export const f32 = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : s;  
};
/** Format `number` as WGSL `u32` */
export const u32 = (x: number) => Math.round(x).toString() + 'u';
/** Format `number` as WGSL `i32` */
export const i32 = (x: number) => Math.round(x).toString();
