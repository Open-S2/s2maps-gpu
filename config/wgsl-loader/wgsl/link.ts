import { Tree } from '@lezer/common';
import { ShaderDefine } from '../types';

import { defineConstants, defineEnables, loadModule, loadModuleWithCache, DEFAULT_CACHE } from './shader';
import { rewriteUsingAST } from './ast';
import { makeLinker, makeLinkCode, makeLinkBundle, makeLinkModule } from '../util/link';

// No preamble
const getPreambles = () => [];

// Allow attribute renaming
const getRenames = (
  defines?: Record<string, ShaderDefine> | null,
) => {
  const rename = new Map<string, string>();
  if (defines) for (let k in defines) if (k[0] === '@') rename.set(k, `${defines[k]}`);
  return rename;
}

export const linker     = makeLinker(getPreambles, getRenames, defineConstants, defineEnables, rewriteUsingAST);
export const linkBundle = makeLinkBundle(linker);
export const linkModule = makeLinkModule(linker);
export const linkCode   = makeLinkCode(linker, loadModuleWithCache, DEFAULT_CACHE);
