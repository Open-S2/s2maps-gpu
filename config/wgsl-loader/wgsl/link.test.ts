import { WGSLModules } from './wgsl.test.data';
import { linkCode, linkModule } from './link';
import { loadModule, wgsl } from './shader';
import { formatAST } from '../util/tree'; 
import { addASTSerializer } from '../test/snapshot';
import mapValues from 'lodash/mapValues';

const loadedModules = mapValues(WGSLModules, (v, k) => loadModule(v, k, k));

addASTSerializer(expect);

describe("link", () => {
  
  it("links an external", () => {
    
    const code = `
    @link fn getColor() -> vec4<f32> {};
    fn main() -> @location(0) vec4<f32> {
      return getColor();
    }
    `
    
    const getColor = `
    @export fn main() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    `
    
    const linked = linkCode(code, {}, {getColor});
    expect(linked).toMatchSnapshot();

  });

  it("substitutes attributes", () => {
    
    const code = `
    @group(GROUP) @binding(BINDING) var<uniform> color: vec4<f32>;

    @link fn getColor() -> vec4<f32> {};
    fn main() -> @location(LOCATION) vec4<f32> {
      return getColor() + color;
    }
    `
    
    const getColor = `
    @export fn getColor() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    `
    
    const defs = {
      '@group(GROUP)': '@group(0)',
      '@binding(BINDING)': '@binding(0)',
      '@location(LOCATION)': '@location(0)',
    };
    const linked = linkCode(code, {}, {getColor}, defs);
    expect(linked).toMatchSnapshot();

  });

  it("links quad vertex", () => {

    const getPosition = wgsl`
    @export fn getPosition(index: i32) -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    `

    const getPerspective = wgsl`
    @export fn getPerspective(index: i32) -> f32 { return 1.0; }
    `

    const getColor = wgsl`
    @export fn getColor(index: i32) -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    `

    const getSize = wgsl`
    @export fn getSize(index: i32) -> f32 { return 1.0; }
    `

    const getDepth = wgsl`
    @export fn getDepth(index: i32) -> f32 { return 0.5; }
    `

    const module = loadedModules['getQuadVertex'];
    const linked = linkModule(module, loadedModules, {getPosition, getPerspective, getColor, getSize, getDepth});
    expect(linked).toMatchSnapshot();

  });
  
  it("lifts recursive dependency", () => {
    
    const code = `
    use 'getLifted'::{ getLifted };
    use 'getColor1'::{ getColor1 };

    fn main() -> vec4<f32> {
      return getColor1();
    }
    `

    const getLifted = `
    // Lifted Code
    @export fn getLifted() {};
    `

    const getColor1 = `
    use 'getColor2'::{ getColor2 };
    @export fn getColor1() -> vec4<f32> { return getColor2(); }
    `
    
    const getColor2 = `
    use 'getLifted'::{ getLifted };
    @export fn getColor2() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    `
    
    const linked = linkCode(code, {getColor1, getColor2, getLifted});
    expect(linked.indexOf('// Lifted Code')).toBeLessThan(linked.indexOf('getColor2'));
    expect(linked).toMatchSnapshot();

  });
  
  it("tree shakes constants", () => {
    const sub = `
    const colorUsed = vec4<f32>(0.0, 0.1, 0.2, 0.0);
    const colorNotUsed = vec4<f32>(0.0, 0.1, 0.2, 1.0);

    @export fn getColor() -> vec4<f32> {
      return colorUsed;
    }
    `

    const main = `
    @link fn getColor() -> vec4<f32> {};
    fn main() {
      var a = getColor();
    }
    `

    for (let compressed of [false, true]) {
      const modMain = loadModule(main, 'main', undefined, compressed);
      const modSub = loadModule(sub, 'sub', undefined, compressed);

      const getPosition = {...modSub, entry: 'getPosition'};
      const getColor = {...modSub, entry: 'getColor'};

      const linked = linkModule(modMain, {}, {getPosition, getColor});
      expect(linked).toMatchSnapshot();
    }
  })
  
  it("tree shakes around identifiers", () => {

    const sub = `
    fn used() -> f32 { return 1.0; }

    fn unused() -> f32 { return 1.0; }
    
    @export fn getPosition(index: i32) -> vec4<f32> { return vec4<f32>(used(), 0.0, 1.0, 1.0); }

    @export fn getColor(index: i32) -> vec4<f32> { return vec4<f32>(used(), 0.0, 1.0, 1.0); }
    `

    const main = `
    @link fn getPosition(index: i32) -> vec4<f32> {};
    fn main() {
      var a = getPosition(0);
    }
    `

    for (let compressed of [false, true]) {
      const modMain = loadModule(main, 'main', undefined, compressed);
      const modSub = loadModule(sub, 'sub', undefined, compressed);

      const getPosition = {...modSub, entry: 'getPosition'};
      const getColor = {...modSub, entry: 'getColor'};

      const linked = linkModule(modMain, {}, {getPosition, getColor});
      expect(linked).toMatchSnapshot();
    }

  });
  
  it("links same module twice with different entry point", () => {

    const sub = `
    fn used() -> f32 { return 1.0; }
    
    @export fn getPosition(index: i32) -> vec4<f32> { return vec4<f32>(used(), 0.0, 1.0, 1.0); }

    @export fn getColor(index: i32) -> vec4<f32> { return vec4<f32>(used(), 0.0, 1.0, 1.0); }
    `

    const main = `
    @link fn getPosition(index: i32) -> vec4<f32> {};
    @link fn getColor(index: i32) -> vec4<f32> {};
    fn main() {
      var a = getPosition(0);
      var b = getColor(0);
    }
    `

    const modMain = loadModule(main, 'main', undefined, false);
    const modSub = loadModule(sub, 'sub', undefined, false);

    const getPosition = {...modSub, entry: 'getPosition'};
    const getColor = {...modSub, entry: 'getColor'};

    const linked = linkModule(modMain, {}, {getPosition, getColor});
    expect(linked).toMatchSnapshot();

  });

  it("links a global across a module", () => {

    const sub1 = `
    @global @export fn getPosition(index: i32) -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    `

    const sub2 = `
    fn getPosition() {};
    
    @export fn getColor(index: i32) -> vec4<f32> {
      getPosition();
      return vec4<f32>(1.0, 0.0, 1.0, 1.0);
    }
    `

    const main = `
    @link fn getPosition(index: i32) -> vec4<f32> {};
    @link fn getColor(index: i32) -> vec4<f32> {};
    fn main() {
      var a = getPosition(0);
    }
    `

    const modMain = loadModule(main, 'main');
    const modSub1 = loadModule(sub1, 'sub1');
    const modSub2 = loadModule(sub2, 'sub2');

    const getPosition = {...modSub1, entry: 'getPosition'};
    const getColor = {...modSub2, entry: 'getColor'};

    const linked = linkModule(modMain, {}, {getPosition, getColor});
    expect(linked).toMatchSnapshot();

    expect(linked).toMatch(/fn getPosition\(index: i32\)/);
    expect(linked).not.toMatch(/fn _[A-Za-z0-9]{2,}_getPosition\(index: i32\)/);

    expect(linked).not.toMatch(/fn getPosition\(\)/);
    expect(linked).toMatch(/fn _[A-Za-z0-9]{2,}_getPosition\(\)/);

  });
  
  it("links a struct and field", () => {

    const sub1 = `
    @export struct Foo { bar: f32 };
    @export fn access(foo: Foo) -> f32 { return foo.bar; }
    fn bar() -> Foo { return Foo(1.0); }
    `

    const main = `
    @link struct Foo;
    @link fn access(foo: Foo) -> f32;

    fn main() {
      var a = getPosition(0);
    }
    `

    const modMain = loadModule(main, 'main');
    const modSub1 = loadModule(sub1, 'sub1');

    const Foo = {...modSub1, entry: 'Foo'};
    const access = {...modSub1, entry: 'access'};

    const linked = linkModule(modMain, {}, {Foo, access});
    expect(linked).toMatchSnapshot();

    expect(linked).toMatch(/Foo { bar: f32 }/);
    expect(linked).not.toMatch(/fn _[A-Za-z0-9]{2,}bar\(\)/);

  });

  it("hoists an enable directive", () => {
    
    const code = `
    @link fn getColor() -> vec4<f32> {};
    fn main() -> @location(0) vec4<f32> {
      return getColor();
    }
    `
    
    const getColor = `
    enable f16;
    @export fn main() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    `
    
    const linked = linkCode(code, {}, {getColor});
    expect(linked).toMatchSnapshot();

  });
  
});
