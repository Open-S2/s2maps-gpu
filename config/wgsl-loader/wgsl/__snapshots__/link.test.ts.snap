// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link hoists an enable directive 1`] = `
"enable f16;

    fn _01_main() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    fn main() -> @location(0) vec4<f32> {
      return _01_main();
    }
    "
`;

exports[`link lifts recursive dependency 1`] = `
"
    // Lifted Code
    fn _01_getLifted() {};
    

    fn _02_getColor2() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    fn _03_getColor1() -> vec4<f32> { return _02_getColor2(); }
    

    fn main() -> vec4<f32> {
      return _03_getColor1();
    }
    "
`;

exports[`link links a global across a module 1`] = `
"
    fn getPosition(index: i32) -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    fn _02_getPosition() {};
    
    fn _02_getColor(index: i32) -> vec4<f32> {
      _02_getPosition();
      return vec4<f32>(1.0, 0.0, 1.0, 1.0);
    }
    

    fn main() {
      var a = getPosition(0);
    }
    "
`;

exports[`link links a struct and field 1`] = `
"
    struct _01_Foo { bar: f32 };
    fn _01_access(foo: _01_Foo) -> f32 { return foo.bar; }


    fn main() {
      var a = getPosition(0);
    }
    "
`;

exports[`link links an external 1`] = `
"
    fn _01_main() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    fn main() -> @location(0) vec4<f32> {
      return _01_main();
    }
    "
`;

exports[`link links quad vertex 1`] = `
"

const _01_QUAD: array<vec2<i32>, 4> = array<vec2<i32>, 4>(
  vec2<i32>(0, 0),
  vec2<i32>(1, 0),
  vec2<i32>(0, 1),
  vec2<i32>(1, 1),
);

fn _01_getQuadIndex(vertex: i32) -> vec2<i32> {
  return _01_QUAD[vertex];
}

fn _01_getQuadUV(vertex: i32) -> vec2<f32> {
  return vec2<f32>(_01_getQuadIndex(vertex));
}



struct _02_SolidVertex {
  position: vec4<f32>,
  color: vec4<f32>,
  uv: vec2<f32>,
};



struct _03_ViewUniforms {
  projectionMatrix: mat4x4<f32>,
  viewMatrix: mat4x4<f32>,
  viewPosition: vec4<f32>,
  viewResolution: vec2<f32>,
  viewSize: vec2<f32>,
  viewWorldDepth: f32,
  viewPixelRatio: f32,
};

@group(VIEW) @binding(VIEW) var<uniform> _03_viewUniforms: _03_ViewUniforms;

fn _03_worldToView(position: vec4<f32>) -> vec4<f32> {
  return _03_viewUniforms.viewMatrix * position;
}

fn _03_viewToClip(position: vec4<f32>) -> vec4<f32> {
  return _03_viewUniforms.projectionMatrix * position;
}

fn _03_worldToClip(position: vec4<f32>) -> vec4<f32> {
  return _03_viewToClip(_03_worldToView(position));
}

fn _03_getPerspectiveScale(w: f32, f: f32) -> f32 {
  var m = _03_viewUniforms.projectionMatrix;
  var worldScale = m[1][1] * _03_viewUniforms.viewWorldDepth;
  var clipScale = mix(1.0, worldScale / w, f);
  var pixelScale = clipScale * _03_viewUniforms.viewPixelRatio;
  return pixelScale;
}


    fn _04_getPosition(index: i32) -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    fn _05_getColor(index: i32) -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    fn _06_getSize(index: i32) -> f32 { return 1.0; }
    

    fn _07_getDepth(index: i32) -> f32 { return 0.5; }
    

fn getQuadVertex(vertexIndex: i32, instanceIndex: i32) -> _02_SolidVertex {
  var position = _04_getPosition(instanceIndex);
  var color = _05_getColor(instanceIndex);
  var size = _06_getSize(instanceIndex);
  var depth = _07_getDepth(instanceIndex);

  var center = _03_worldToClip(position);

  var uv = _01_getQuadUV(vertexIndex);
  var xy = uv * 2.0 - 1.0;

  // Lerp between fixed size and full perspective.
  var pixelScale = _03_getPerspectiveScale(center.w, depth);
  // TODO: awaiting compound support
  //size *= pixelScale;
  size = size * pixelScale;

  if (HAS_EDGE_BLEED) {
    xy = xy * (size + 0.5) / size;
    uv = xy * .5 + .5;
  }

  // TODO: awaiting compound support
  //center.xy += xy * size * viewUniforms.viewResolution * center.w;
  center = vec4<f32>(center.xy + xy * size * _03_viewUniforms.viewResolution * center.w, center.zw);

  return _02_SolidVertex(
    center,
    color,
    uv
  );
}
"
`;

exports[`link links same module twice with different entry point 1`] = `
"
    fn _01_used() -> f32 { return 1.0; }
    
    fn _01_getPosition(index: i32) -> vec4<f32> { return vec4<f32>(_01_used(), 0.0, 1.0, 1.0); }

    fn _01_getColor(index: i32) -> vec4<f32> { return vec4<f32>(_01_used(), 0.0, 1.0, 1.0); }
    

    fn main() {
      var a = _01_getPosition(0);
      var b = _01_getColor(0);
    }
    "
`;

exports[`link substitutes attributes 1`] = `
"
    fn _01_getColor() -> vec4<f32> { return vec4<f32>(1.0, 0.0, 1.0, 1.0); }
    

    @group(0) @binding(0) var<uniform> color: vec4<f32>;

    fn main() -> @location(0) vec4<f32> {
      return _01_getColor() + color;
    }
    "
`;

exports[`link tree shakes around identifiers 1`] = `
"
    fn _01_used() -> f32 { return 1.0; }
fn _01_getPosition(index: i32) -> vec4<f32> { return vec4<f32>(_01_used(), 0.0, 1.0, 1.0); }

    

    fn main() {
      var a = _01_getPosition(0);
    }
    "
`;

exports[`link tree shakes around identifiers 2`] = `
"
    fn _01_used() -> f32 { return 1.0; }
fn _01_getPosition(index: i32) -> vec4<f32> { return vec4<f32>(_01_used(), 0.0, 1.0, 1.0); }

    

    fn main() {
      var a = _01_getPosition(0);
    }
    "
`;

exports[`link tree shakes constants 1`] = `
"
    const _01_colorUsed = vec4<f32>(0.0, 0.1, 0.2, 0.0);
fn _01_getColor() -> vec4<f32> {
      return _01_colorUsed;
    }
    

    fn main() {
      var a = _01_getColor();
    }
    "
`;

exports[`link tree shakes constants 2`] = `
"
    const _01_colorUsed = vec4<f32>(0.0, 0.1, 0.2, 0.0);
fn _01_getColor() -> vec4<f32> {
      return _01_colorUsed;
    }
    

    fn main() {
      var a = _01_getColor();
    }
    "
`;
